src/server/services/wfs3/qgswfs3handlers.cpp

```diff
@@ -1395,23 +1399,94 @@ void QgsWfs3CollectionsItemsHandler::handleRequest( const QgsServerApiContext &c
-        prevLink["rel"] = "prev";
-        prevLink["name"] = "Previous page";
-        data["links"].push_back( prevLink );
-      }
+      // Pagination metadata
+      json pagination = json::array();
 
-      if ( limit + offset < matchedFeaturesCount )
+      if ( limit != 0 )
       {
-        json nextLink = selfLink;
-        nextLink["href"] = QStringLiteral( "%1&offset=%2&limit=%3" ).arg( cleanedUrlAsString ).arg( std::min<long>( matchedFeaturesCount, limit + offset ) ).arg( limit ).toStdString();
-        nextLink["rel"] = "next";
-        nextLink["name"] = "Next page";
-        data["links"].push_back( nextLink );
+        // Add prev - next links
+        json prevLink;
+        if ( offset != 0 )
+        {
+          prevLink = selfLink;
+          prevLink["href"] = cleanedUrlAsString.toStdString() + QStringLiteral( "offset=%1&limit=%2" ).arg( std::max<long>( 0, offset - limit ) ).arg( limit ).toStdString();
+          prevLink["rel"] = "prev";
+          prevLink["title"] = "Previous page";
+          data["links"].push_back( prevLink );
+        }
+
+        json nextLink;
+        if ( limit + offset < matchedFeaturesCount )
+        {
+          nextLink = selfLink;
+          nextLink["href"] = cleanedUrlAsString.toStdString() + QStringLiteral( "offset=%1&limit=%2" ).arg( std::min<long>( matchedFeaturesCount, limit + offset ) ).arg( limit ).toStdString();
+          nextLink["rel"] = "next";
+          nextLink["title"] = "Next page";
+          data["links"].push_back( nextLink );
+        }
+
+        // Pagination
+        if ( matchedFeaturesCount - limit > 0 )
+        {
+          const int totalPages { static_cast<int>( std::ceil( static_cast<float>( matchedFeaturesCount ) / static_cast<float>( limit ) ) ) };
+          const int currentPage { static_cast<int>( offset / limit + 1 ) };
+          const std::string currentPageLink { selfLink["href"] };
+
+          std::string prevPageLink;
+          if ( prevLink.contains( std::string{ "href" } ) )
+          {
+            prevPageLink = prevLink["href"];
+          }
+
+          std::string nextPageLink;
+          if ( nextLink.contains( std::string{ "href" } ) )
+          {
+            nextPageLink = nextLink["href"];
+          }
+
+          const std::string firstPageLink { cleanedUrlAsString.toStdString() + QStringLiteral( "offset=0&limit=%1" ).arg( limit ).toStdString() };
+          const std::string lastPageLink { cleanedUrlAsString.toStdString() + QStringLiteral( "offset=%1&limit=%2" ).arg( totalPages * limit - limit ).arg( limit ).toStdString() };
+
+          if ( currentPage != 1 )
+          {
+            pagination.push_back( {{ "title",  "1" }, { "href", firstPageLink }, { "class", "page-item" }} ) ;
+          }
+          if ( currentPage > 3 )
+          {
+            pagination.push_back( {{ "title", "..." }, { "class", "page-item disabled" }} ) ;
```

Comment:
Maybe use the unicode char for three dots: …

Reply:
Good idea, addressed per https://github.com/qgis/QGIS/pull/52051/commits/8fc94611895dca931867e6748971c2bb69754fa4